File: Algorithm/LeetCode/T0239-sliding-window-maximum.py
from typing import List
import heapq

# https://leetcode-cn.com/problems/sliding-window-maximum/
# 困难，优先队列
def maxSlidingWindow(nums: List[int], k: int) -> List[int]:
  # 注意 Python 默认的优先队列是小根堆
  q = [(-nums[i], i) for i in range(k)]
  heapq.heapify(q)

  ans = [-q[0][0]]
  for i in range(k, len(nums)):
      heapq.heappush(q, (-nums[i], i))
      while q[0][1] <= i - k:
          heapq.heappop(q)
      ans.append(-q[0][0])
  
  return ans

File: Algorithm/LeetCode/T0239-sliding-window-maximum.js
/**
 * https://leetcode-cn.com/problems/sliding-window-maximum/
 * 困难，单调队列
 * TODO: 复习
 * @param {number[]} nums
 * @param {number} k
 * @return {number[]}
 */
var maxSlidingWindow = function (nums, k) {
  if (k === 1) {
    return nums;
  }

  const q = [];

  for (let i = 0; i < k; i++) {
    while (q.length && nums[i] >= nums[q[q.length - 1]]) {
      q.pop();
    }
    q.push(i);
  }

  const result = [nums[q[0]]];
  for (let i = k; i < nums.length; i++) {
    while (q.length && nums[i] >= nums[q[q.length - 1]]) {
      q.pop();
    }
    q.push(i);

    while (q[0] <= i - k) {
      q.shift();
    }

    result.push(nums[q[0]]);
  }

  return result;
};
